plugins/destination/plugin_testing_migrate.go

```diff
@@ -3,124 +3,281 @@ package destination
 	}
 
-	sourceName := tableName
+	sourceName := target.Name
 	sourceSpec := specs.Source{
 		Name: sourceName,
 	}
 	syncTime := time.Now().UTC().Round(1 * time.Second)
-	resource1 := createTestResources(table, sourceName, syncTime, 1)[0]
-	if err := p.writeOne(ctx, sourceSpec, []*schema.Table{table}, syncTime, resource1); err != nil {
+	resource1 := createTestResources(source, sourceName, syncTime, 1)[0]
+	if err := p.writeOne(ctx, sourceSpec, []*schema.Table{source}, syncTime, resource1); err != nil {
 		return fmt.Errorf("failed to write one: %w", err)
 	}
 
-	// check that migrations and writes still succeed when column ordering is changed
-	a := table.Columns.Index("uuid")
-	b := table.Columns.Index("float")
-	table.Columns[a], table.Columns[b] = table.Columns[b], table.Columns[a]
-	if err := p.Migrate(ctx, []*schema.Table{table}); err != nil {
-		return fmt.Errorf("failed to migrate table with changed column ordering: %w", err)
+	if err := p.Migrate(ctx, []*schema.Table{target}); err != nil {
+		return fmt.Errorf("failed to migrate existing table: %w", err)
 	}
-	resource2 := createTestResources(table, sourceName, syncTime, 1)[0]
-	if err := p.writeOne(ctx, sourceSpec, []*schema.Table{table}, syncTime, resource2); err != nil {
-		return fmt.Errorf("failed to write one after column order change: %w", err)
+	resource2 := createTestResources(target, sourceName, syncTime, 1)[0]
+	if err := p.writeOne(ctx, sourceSpec, []*schema.Table{target}, syncTime, resource2); err != nil {
+		return fmt.Errorf("failed to write one after migration: %w", err)
 	}
 
-	resourcesRead, err := p.readAll(ctx, table, sourceName)
+	resourcesRead, err := p.readAll(ctx, target, sourceName)
 	if err != nil {
 		return fmt.Errorf("failed to read all: %w", err)
 	}
-	if len(resourcesRead) != 2 {
-		return fmt.Errorf("expected 2 resources after second write, got %d", len(resourcesRead))
+	if mode == specs.MigrateModeSafe {
+		if len(resourcesRead) != 2 {
+			return fmt.Errorf("expected 2 resources after write, got %d", len(resourcesRead))
+		}
+		if diff := resourcesRead[1].Diff(resource2.Data); diff != "" {
+			return fmt.Errorf("resource2 diff: %s", diff)
+		}
+	} else {
+		if len(resourcesRead) != 1 {
+			return fmt.Errorf("expected 1 resources after write, got %d", len(resourcesRead))
+		}
+		if diff := resourcesRead[0].Diff(resource2.Data); diff != "" {
+			return fmt.Errorf("resource1 diff: %s", diff)
+		}
 	}
 
-	// check that migrations succeed when a new column is added
-	table.Columns = append(table.Columns, schema.Column{
-		Name: "new_column",
-		Type: schema.TypeInt,
+	return nil
+}
+
+func (*PluginTestSuite) destinationPluginTestMigrate(
+	ctx context.Context,
+	t *testing.T,
+	newPlugin NewPluginFunc,
+	logger zerolog.Logger,
+	spec specs.Destination,
+	strategy MigrateStrategy,
+) {
+	spec.BatchSize = 1
+
+	t.Run("add_column", func(t *testing.T) {
+		if strategy.AddColumn == specs.MigrateModeForced && spec.MigrateMode == specs.MigrateModeSafe {
```

Comment:
I'm a bit confused by this, does this mean "Only run the test if it's expected to pass?". I mean `safe migrate mode declared in test && safe migrate mode declared in spec` or `forced migrate mode declared in test && forced migrate mode declared in spec`?

Don't we also want to test that there's an error when we have `spec safe mode` and can't migrate?

Reply:
I think for not we don't want to check for error yet as there are a lot of tests already (this will be `add_column_error` test varient). so I would give it some time to see if this approach is at all valid and then we can improve further.